compiler/rustc_resolve/src/late.rs

```diff
@@ -1318,7 +1321,12 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
                             visit::walk_trait_ref(this, trait_ref);
                         }
                         // Resolve the self type.
+                        let prev = replace(
+                            &mut this.diagnostic_metadata.currently_processing_impl_self_ty,
+                            true,
+                        );
                         this.visit_ty(self_type);
```

Comment:
What happens if we visit `self_type` outside of the call to `with_self_rib(res,`?

Reply:
If we do, `visit_ty` just visits `self_ty` unless `currently_processing_impl_self_ty` is true. It never causes a `Self is only available in impls, traits, and type definitions` error.

In reference to https://github.com/rust-lang/rust/pull/93971#pullrequestreview-881042951, should I improve this error to report it right away?